### DISCUSSION TITLE
How can I get minified css with bundle-text?

### DISCUSSION THREAD

**Discussion by danieltroger** [parcel-bundler/parcel#5751] [Q&A] [answered]
# Preface:
I need parcel to output 1 file.

I have previously, to import css, made a .ts file which does
```
export css = `
.mycss {
goes:here;
}
`;
```

And then imported that and added it to a `<style>` element in my javascript.

However as my css has gotten more and more there are like 1000 `\n`'s in my output code and lke 20 times as many whitespaces (because of the indentation).

I then moved to importing the css with fs.readFileSync(<file>, "utf8"); which is nice because I now at least get syntax highlighting when writing the css, but doesn't solve the minification issue.

Now I wanted to ask how to use the css pipeline but get the result inlined into the js instead, but first I read the docs and some issues.

There's a heap of issues around this and I gave up on trying to understand them all. But my takeaway was that a simpler way of the readFileSync thing was added, ["bundle-text"](https://v2.parceljs.org/configuration/plugin-configuration/#predefined-(offical)-named-pipelines)

# The question

I tried out bundle-text and it's nice, but the css in the transpiled file looks exactly the same as when using fs.readFileSync. Is there some way to get it minified? If so, how? I would like to avoid having to preprocess my stuff before calling parcel.

I tried renaming it to .less and running `npm install --save-dev less` but it still looks the same.

**Comment by mischnic** [comment, marked as answer]
> I tried out bundle-text and it's nice, but the css in the transpiled file looks exactly the same as when using fs.readFileSync. Is there some way to get it minified? If so, how? I would like to avoid having to preprocess my stuff before calling parcel.

if you do a production build (= `parcel build`), it should be minified just like any other css.

> less

The example on the docs is just to illustrate that `bundle-text:` is more powerful than `readFileSync` (because it's actually processed).

**Reply by danieltroger** [reply]
> if you do a production build (= parcel build), it should be minified just like any other css.

I usually build with `--no-minify` (now `--no-optimize`) and minify with `terser -cm toplevel,eval -- "$target.js" > "$minifiedfn"` because the output files are way smaller then what parcel outputs when I let it minify anything containing eval (122KB vs 207KB). Also calling terser with only `-cm` is consistently smaller.

I know there is some way to change parcel's terser options but I haven't had time to figure it out. Do you happen to have a link to some docs how I can configure the terser options?

# Regarding the css

It didn't help to build without `--no-optimize`. There are still `\n`'s and plenty of whitespaces.

I also found this: http://lesscss.org/usage/#less-options and this https://v2.parceljs.org/languages/less and tried this
```
$ cat > lessrc.js 
{
  "compress": true
}
```

But nothing changed. And I found this https://github.com/parcel-bundler/parcel/discussions/4708 and tried to rename my css to .pcss and it's still not minified. But I will try .pcss again without `--no-optimize`.

Can you confirm that just importing normal css with bundle-text: should minify it? If that's the case I will try to create a simple reproduction and file it as a bug if it doesn't make me discover what I'm doing wrong.

**Reply by danieltroger** [reply]
yeah it definitely doesn't get processed, I'll open an issue

**Reply by mischnic** [reply]
> terser -cm toplevel,eval

Adding a .terserrc with `{"mangle": {"eval": true}}` should make Parcel use that config as well. (Use the syntax described in the Terser API Reference options: https://github.com/terser/terser#api-reference)

> yeah it definitely doesn't get processed, I'll open an issue

👍

**Reply by danieltroger** [reply]
> Adding a .terserrc with {"mangle": {"eval": true}} should make Parcel use that config as well. (Use the syntax described in the Terser API Reference options: https://github.com/terser/terser#api-reference)

Thanks a lot! There's still a 79-byte difference on 122KB of code. Minifying with terser after parcel seems to do it more intelligently, i.e. variables only used once are not declared but instead the value is just put where it is used, semi colons that aren't needed are removed, if statements are converted to ternary statements.

I will switch when parcel wins due to minified css ;)

Opened #5755

**Reply by mischnic** [reply]
That's very odd. It should be identical. These are the default terser options: https://github.com/parcel-bundler/parcel/blob/ac8220295a96710618b376ae25501e74ea5b5c44/packages/optimizers/terser/src/TerserOptimizer.js#L27-L40

**Reply by danieltroger** [reply]
Maybe parcel is hardcoded to an old terser version? I will come back at this later if I have time. Possibly try to make it reproducible somehow without uploading my entire code if I can't figure out which options differ.

**Reply by mischnic** [reply]
Not really hardcoded, but it might be different depending on your lockfile: https://github.com/parcel-bundler/parcel/blob/1188e59f87ec8d899152932d7c3eec1b18cc2579/packages/optimizers/terser/package.json#L28

**Reply by danieltroger** [reply]
@mischnic I deleted my package-lock.json and terser still produces smaller files.

I managed to dump the minify options from inside terser. The terser code actually running is in `node_modules/terser/dist/bundle.min.js` FYI.

So this is how it looks when running `terser -cm`:

```
{
  compress: true,
  mangle: true,
  ecma: undefined,
  enclose: false,
  ie8: false,
  keep_classnames: undefined,
  keep_fnames: false,
  module: false,
  nameCache: null,
  output: null,
  format: null,
  parse: {},
  rename: undefined,
  safari10: false,
  sourceMap: false,
  timings: false,
  toplevel: false,
  warnings: false,
  wrap: false
}
```

and this is how it looks when using parcel with no .terserrc:
```
{
  sourceMap: false,
  toplevel: false,
  module: false,
  compress: {},
  ecma: undefined,
  enclose: false,
  ie8: false,
  keep_classnames: undefined,
  keep_fnames: false,
  mangle: {},
  nameCache: null,
  output: null,
  format: null,
  parse: {},
  rename: undefined,
  safari10: false,
  timings: false,
  warnings: false,
  wrap: false
}
```

The only difference seems to be that compress and mangle are objects in parcel instead of true. Giving up on this again for now, got stuff to do.

**Reply by danieltroger** [reply]
I think those are the mangle opts?

parcel
```
{
  cache: null,
  eval: false,
  ie8: false,
  keep_classnames: false,
  keep_fnames: false,
  module: false,
  properties: false,
  reserved: [],
  safari10: false,
  toplevel: false
}
```

terser:
```
{
  cache: null,
  eval: false,
  ie8: false,
  keep_classnames: false,
  keep_fnames: false,
  module: false,
  properties: false,
  reserved: [],
  safari10: false,
  toplevel: false
}
```

**Reply by danieltroger** [reply]
@mischnic is it possible to have different .terserrc's per target btw? I want to allow terser to convert function's to arrow functions for the modern target only

**Reply by mischnic** [reply]
> is it possible to have different .terserrc's per target btw?

No. The current terser optimizer doesn't allow this. But plugin API-wise, it's possible.

**Reply by danieltroger** [reply]
@mischnic do you happen to know some easy-to-switch-to alternatives to cssnano?

It even fails at simple deduplication as soon as media queries are involved :( (see issue 990 on their repo).

**Reply by mischnic** [reply]
You might be able to disable some specific problematic cssnano steps.
